components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt

```diff
@@ -108,6 +113,24 @@ class BrowserIcons @Suppress("LongParameterList") constructor(
     private val minimumSize = context.resources.getDimensionPixelSize(R.dimen.mozac_browser_icons_minimum_size)
     private val scope = CoroutineScope(jobDispatcher)
 
+    init {
+        loaders = loaders + if (deferLoadingNetworkIcons) {
+            NonBlockingHttpIconLoader(httpClient, scope) { request, resource, result ->
+                val desiredSize = request.getDesiredSize(context, minimumSize, maximumSize)
+
+                val icon = decodeIconLoaderResult(result, decoders, desiredSize)
+                    ?: generator.generate(context, request)
+
+                process(
+                    context, processors.filterIsInstance<PersistingIconProcessor>(),
+                    request, resource, icon, desiredSize
+                )
+            }
```

Comment:
We are kind of duplicating the pipeline here. Which is hard to understand when reading, and also makes further changes hard. 🤔 

It makes me once again aware that I chose the wrong abstraction here. I wanted us to be able to handle abstract loaders and processors. But that's way too generic and awkward. We want to define a very specific logic and order of things. If this logic/order would be explicit then it would make things like this here way easier to implement (either wait for the network request or not).

I think we should make this change soon. It will also allows us to easily fix things like https://github.com/mozilla-mobile/android-components/issues/10377.

But in the meantime we may want to land a change like this here. The refactoring will be a bit bigger. :)

Reply:
`We are kind of duplicating the pipeline here.`
Indeed. And it may not be obvious. :-( .
Looked into extracting / refactoring the current functionality to then be used in both places but at the same time there are enough differences that I couldn't get something reasonably better.
This is a quick fix but a bigger refactoring seems indeed needed.

Reply:
I agree. I think the duplication is okay since eventually we will change this code.